File: leetcode/hot100/codes/Python/单词拆分.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True

        for i in range(1, n + 1):
            for word in wordDict:
                len_word = len(word)
                if len_word <= i and dp[i] == False:
                    sub_str = s[i - len_word:i]
                    if sub_str == word:
                        dp[i] = dp[i - len_word]

        return dp[n]

File: leetcode/hot100/codes/C++/单词拆分.cpp
class Solution {
    public:
        bool wordBreak(string s, vector<string>& wordDict) {
            int n = s.size();
            vector<bool> dp(n + 1, false);
    
            dp[0] = true;
            for(int i = 1; i <= n; i++){
                for(string word : wordDict){
                    int len_word = word.size();
                    if(len_word <= i && dp[i] == false){
                        string sub_str = s.substr(i - len_word, len_word);
                        if(sub_str == word){
                            dp[i] = dp[i - len_word];
                        } 
                    }
                }
            }
    
            return dp[n];
        }
    };